### ISSUE TITLE
Investigate Meson build system

### ISSUE THREAD

**Issue by shoogle** [AppImage/AppImageKit#205]
Meson is a build system which is designed to solve the problems with existing build tools like make, automake, Cmake, qmake, etc. It is designed to be fast and simple to use. It has an interesting dependency management system that might make it ideally suited for building AppImages - either through scripts or by adding native AppImage support directly to Meson.

https://github.com/mesonbuild/meson
http://mesonbuild.com/
http://www.admin-magazine.com/HPC/Articles/The-Meson-Build-System

Highlights from [Meson presentation video](https://www.youtube.com/watch?v=KPi0AuVpxLI) on YouTube:
- Design goals: https://youtu.be/KPi0AuVpxLI?t=7m4s
- Best features kept from other build systems: https://youtu.be/KPi0AuVpxLI?t=11m57s
- Building a C++ project with Meson: https://youtu.be/KPi0AuVpxLI?t=17m16s
- Building a Qt project with Meson: https://youtu.be/KPi0AuVpxLI?t=20m33s
- Benchmarks (compiling GLib): https://youtu.be/KPi0AuVpxLI?t=21m45s
- Cross compiling: https://youtu.be/KPi0AuVpxLI?t=26m9s
- IDE compatibility: https://youtu.be/KPi0AuVpxLI?t=30m43s
- Distribution/developer library war: https://youtu.be/KPi0AuVpxLI?t=33m4s

Native AppImage support could take the form of an `--AppImage` command line option for Meson which would not only build the application, but would also locate all of the necessary libraries and components and copy them into the build directory (i.e. it would create the AppDir).

**Comment by probonopd** [comment]
Very interesting indeed!

**Comment by shoogle** [comment]
It looks like the Meson developer had a similar idea!

- Creating static/standalone binaries:
  - [guide for linux](https://github.com/mesonbuild/meson/wiki/Creating-Linux-binaries)
  - [examples for all platforms](https://github.com/mesonbuild/meson/tree/master/manual%20tests/4%20standalone%20binaries)
- [Wrap dependency management (all platforms)](
https://github.com/mesonbuild/meson/wiki/Wrap%20dependency%20system%20manual)

Summary:
- Just pass an option and an environment variable to Meson and it will statically link all the libraries it can link statically, and copy in the `lib.so` files for all the libraries that can't be linked statically.
- Obviously, this requires the libraries to be present on the build system already, and this might involve building some of them from source. The Wrap dependency system makes this relatively easy.

**Comment by probonopd** [comment]
Sounds intriguing.

> This directory shows how you can build redistributable binaries. On OSX this menans building an app bundle and a .dmg installer. On Linux it means building a binary that bundles its dependencies. (...) On Linux you must build the package on the oldest distribution you plan to support (Debian stable/oldstable and old CentOS are the common choice here).

I think we should contact the author and ask him whether he would be interested in AppImage as a target. Since he writes "a binary that bundles its dependencies" but actually the script seems to generate an "archive"...

**Comment by shoogle** [comment]
Yes it's probably worth talking to him. I'm really busy at the moment - won't be available until the end of the month - but feel free to start a discussion without me. It would probably be a good idea to have a quick go with Meson first, maybe copy his example project, and see whether his tools can locate and bundle indirect dependencies. The next step would be to implement a blacklist of system libraries.

**Comment by jpakkane** [comment]
Hi, I'm the main Meson developer. Our goal is to make it easily build to any package format, whether it be AppImage, msi installer, dmg or flatpak. I have not looked too deeply into how AppImage organizes itself, but presumably it is mostly a question of setting up paths correctly when doing install and building all dependencies that are not provided by the base system. The former is done with command line arguments to Meson and the latter by having said deps available as wraps.

**Comment by probonopd** [comment]
Hi @jpakkane and welcome to AppImage. I would roughly compare an AppImage to what is an `.app` inside a `.dmg` on the macOS platform. It would be great if you could help by making Meson capable of outputting AppImages.

For an introduction see our [wiki](https://github.com/probonopd/AppImageKit/wiki), especially the part of [how to create portable AppImages](https://github.com/probonopd/AppImageKit/wiki/Creating-AppImages).

When in doubt, consult the [AppImageSpec](https://github.com/AppImage/AppImageSpec).

Key to AppImages that run on not just the most recent but also on older distributions is that the part of the dependency graph which is not supposed to be bundled, starting from glibc, is not more recent than what can be expected to be part of the target system(s). I recommend library versions from the debian oldstable or CentOS 6 era for this reason. Of course, those dependencies that will be bundled inside the AppImage may be more recent than that. A rule-of-thumb for what should be bundled: "Everything that cannot be reasonably expected to be part of each target system in a recent enough version" - namely, everything except [these](https://github.com/probonopd/AppImages/blob/master/excludelist) libraries.

Please let me know here if something is unclear, doesn't make sense, or if you get stuck.

Again, welcome!

**Comment by jpakkane** [comment]
From what I can tell of the documentation, all a project needs to do to support AppImages is to install the AppRun shell script and the PNG image (+ maybe help and the rest). This is quite straightforward and requires no changes in Meson, it can stage an appdir just fine.

**Comment by shoogle** [comment]
@jpakkane Thanks for checking in on AppImages!

Your assessment is pretty accurate. AppImageKit is a very capable tool for making AppImages, and making a ***(very basic)*** AppImage is just a case of adding the AppRun script, icon and .desktop file.

There's no changes needed to Meson there, but here is where you might be able to help us:

#### 1. Detect (and bundle) indirect dependencies:
Where it gets tricky is making the AppImage truly portable, as this involves bundling not only the program's direct dependencies, but also it's dependencies' dependencies and so on. It is already possible to detect these dependencies using `ldd`, but copying them into the container currently requires [writing a pretty complicated script](https://github.com/musescore/MuseScore/blob/master/build/Linux%2BBSD/portable/copy-libs). It would be great if Meson could do this automatically.

#### 2. Detect (and bundle) runtime dependencies:
A major problem is where some dependencies are loaded dynamically at runtime and are not detectable using the usual `ldd` tricks. This is true in particular for many Qt plugins and QML. Since Meson appears to be "aware" of Qt and it's various components it would also be great if it was "aware" of their runtime dependencies and could also copy those in. Again, these must currently be specified manually by running the app and testing features until something breaks, adding the missing library (and its dependencies), and then [trying again](https://github.com/musescore/MuseScore/blob/master/build/Linux%2BBSD/portable/copy-libs#L167), and [again](https://github.com/musescore/MuseScore/blob/master/build/Linux%2BBSD/portable/copy-libs#L193), and then [doing the same on every other distribution](https://github.com/musescore/MuseScore/blob/master/build/Linux%2BBSD/portable/copy-libs#L243).

#### 2. Excluding dependencies provided by the  base-system
It is desirable to exclude all libraries that can be guaranteed present on any Linux system to prevent the AppImage becoming too large. Since there is no standard Linux base system this means developers currently have to use trail and error to see what they can get away with not including.

I think that (1) can be automated within Meson, but (2) and (3) probably require lists to be manually defined. It would be great to be able to have these lists defined centrally somewhere and then have Meson use them to determine what needs to be bundled. The lists would look something like this:

```bash
# Qt Plugins dependencies:
libfoo
libbar
libbaz
...
```
```bash
# Linux base system:
libfoo
libbar
libbaz
...
```
If Meson was used to package an application that depends on Qt Plugins then it would include everything on the first list and exclude everything on the second list.

**Comment by probonopd** [comment]
> It would be great to be able to have these lists defined centrally somewhere

For the list of libraries that are expected to come with the base system and hence should not be bundled, [this](https://github.com/probonopd/AppImages/blob/master/excludelist) list can be used.

**Comment by probonopd** [comment]
Whoever does the work decides which tools to use 💯

**Comment by TheAssassin** [comment]
I'd go for CMake (see #408). Most IDEs and other tool sets support it, it's widely spread and well tested. Although it might not be as elegant as meson, it's probably the better decision to go for CMake.

**Comment by TheAssassin** [comment]
Transitioning to CMake in the near future. Partial implementation is available already. Closing this issue.

**Comment by probonopd** [comment]
Why I absolutely dislike Meson:
https://github.com/openzim/libzim/issues/212

**Comment by shoogle** [comment]
> Why I absolutely dislike Meson: openzim/libzim#212

*TLDR: Meson appears not to work on Ubuntu 14.04 LTS despite the fact that 14.04 is still supported by Canonical (or was until the 25th of April this year). This is apparantly because Meson depends on a newer version of Python than is provided for that distribution.*

However, Meson's [CI tests](https://travis-ci.org/mesonbuild/meson) run on Ubuntu 14.04, so the above conclusion may not be entirely accurate. Anyway, Ubuntu 16.04 is now the oldest supported Ubuntu LTS, and there are [good reasons](https://github.com/AppImage/AppImageKit/issues/166#issuecomment-450728637) to consider switching to it.

**Comment by jpakkane** [comment]
> However, Meson's CI tests run on Ubuntu 14.04

This is not entirely accurate. The CI uses a Docker image that is based on Ubuntu 18/10.

**Comment by shoogle** [comment]
> The CI uses a Docker image that is based on Ubuntu 18/10.

Thanks for the clarification @jpakkane. Perhaps you would consider switching to something a bit older in future because most people use whatever Travis currently uses.

**Comment by probonopd** [comment]
Hi @jpakkane. In this project, we care about building applications for the oldest still-supported long term stable distributions, e.g., CentOS 6, Ubuntu LTS, and so on. Hence, to make it feasible to use Meson for this purpose, it would be good if Meson was developed on say, CentOS 6 and accept no dependencies that are newer than what is shipped with it. 

I realize that our demands are a bit special, so if building on the oldest still-supported long term stable distributions is not in the scope of Meson, then Meson is not a good fit for our purpose but may still be fine for other use cases.

**Comment by TheAssassin** [comment]
Well, meson as a build system wouldn't be problematic strictly in terms of dependencies; we can build additional ones. But then again, that's annoying work to do. So I agree with @probonopd.

**Comment by probonopd** [comment]
That being said, an AppImage that would contain Python, Meson, and ninja and that could run on all still-supported LTS distributions might be beneficial for the Meson project to have - it would allow users to use Meson without having to worry about incompatible Python versions and such, which currently makes using it cumbersome.

**Comment by shoogle** [comment]
As a build system, Meson should really target the oldest distribution possible. This should be by default, for CI Tests and everything, not just as an optional AppImage build for those that want it. Application developers sometimes have to choose between supporting older systems or gaining access to features only provided on newer systems, but they don't expect this decision to be made for them by their build system.

While Meson builds against Ubuntu 18.10 there is a genuine risk that it might not work with Ubuntu 18.04, which is the latest LTS. Most users and developers use LTS releases, so at the very least Meson should support the latest LTS, and preferably the one before that too.

**Comment by jpakkane** [comment]
We do our best to support things, but our test matrix is really big. Running tests is already fairly slow. Adding even more platforms will quickly become unfeasible. The only "real" dependency we have is Python 3 and we do have 3.5 in our test setup (on Windows IIRC, but it's there nevertheless). Thus far we have not received many (or any?) bug reports about things breaking in old LTS versions.

The version of Python required is the oldest that is available from the system in Debian Stable, latest Ubuntu LTS and in the future probably also the latest RHEL.

**Comment by probonopd** [comment]
What we are asking for is to go for the _oldest_ rather than the _newest_ LTS. It should not increase the test matrix because what runs on old systems is supposed to also run on newer systems.

**Comment by shoogle** [comment]
In other words, if you can only support one platform, pick the oldest one and you get the newer ones for free. That's what Qt and everybody else does.

**Comment by Calinou** [comment]
> In other words, if you can only support one platform, pick the oldest one and you get the newer ones for free.

Newer Python versions can sometimes break in subtle ways, so this isn't *always* true. That said, if you have to choose between testing on an old and a new version, it's probably better to test on the old version.

**Comment by shoogle** [comment]
> Newer Python versions can sometimes break in subtle ways

That's true, but that's arguably a Python bug so not something you'd have to worry about. In any case, you can rely on experienced users to test and fix bugs on new systems whereas people on older systems tend to be less experienced (though not always), and the bugs on newer systems are likely to be smaller and easier to fix than if a dependency is completely missing on an old system.
